rmf_fleet_adapter_python/scripts/rmf_fleet_adapter_python/test_adapter.py

```diff
@@ -482,132 +577,72 @@ def updater_inserter(handle_obj, updater):
                     starts,
                     partial(updater_inserter, robot_cmd))
 
-    # INIT TASK OBSERVER ======================================================
-    at_least_one_incomplete = False
-    completed = False
-
-    def task_cb(msg):
-        nonlocal at_least_one_incomplete
-        nonlocal completed
-
-        if msg.state == TASK_STATE_COMPLETED:
-            completed = True
-            at_least_one_incomplete = False
-        else:
-            completed = False
-            at_least_one_incomplete = True
-
-    task_node = rclpy.create_node("task_summary_node")
-    task_node.create_subscription(TaskSummary,
-                                  "task_summaries",
-                                  task_cb,
-                                  10)
-
     # INIT DISPENSERS =========================================================
-    quiet_dispenser = MockQuietDispenser(quiet_dispenser_name)
-    flaky_dispenser = MockFlakyDispenser(flaky_dispenser_name)
+    dispenser = MockDispenser(dispenser_name)
+    ingestor = MockIngestor(ingestor_name)
+
+    # INIT TASK SUMMARY OBSERVER ==============================================
+    observer = TaskSummaryObserver()
 
     # FINAL PREP ==============================================================
     rclpy_executor = SingleThreadedExecutor()
-    rclpy_executor.add_node(task_node)
     rclpy_executor.add_node(cmd_node)
-    rclpy_executor.add_node(quiet_dispenser)
-    rclpy_executor.add_node(flaky_dispenser)
-
-    last_quiet_state = False
-    last_flaky_state = False
+    rclpy_executor.add_node(dispenser)
+    rclpy_executor.add_node(ingestor)
+    rclpy_executor.add_node(observer)
 
     # GO! =====================================================================
     adapter.start()
 
     print("# SENDING NEW REQUEST ############################################")
-    request = adpt.type.CPPDeliveryMsg("test_delivery",
+    test_name = 'test_delivery'
+    request = adpt.type.CPPDeliveryMsg(test_name,
                                        pickup_name,
-                                       quiet_dispenser_name,
+                                       dispenser_name,
                                        dropoff_name,
-                                       flaky_dispenser_name)
-    quiet_dispenser.reset()
-    flaky_dispenser.reset()
+                                       ingestor_name)
```

Comment:
Would wish to have another request to test, since we have a new helpful `TaskSummaryObserver` class. We could try to add a `LoopRequest` task as a test case.

Since the binding is not avail for `request_loop`, might wanna trouble you to add this [here](https://github.com/osrf/rmf_fleet_adapter_python/blob/efa4c1f181ee8ae82fb59e532ab83eb9bec18cea/rmf_fleet_adapter_python/src/adapter.cpp#L167)
```
.def("request_loop", &agv::test::MockAdapter::request_loop)
```

This will definitely result in a more comprehensive integration test. :smile:

Reply:
I've added the necessary bindings for Loop, and added a test for it as well, which checks that the robot only traverses specific vertices. [see here](https://github.com/osrf/rmf_fleet_adapter_python/pull/6/commits/2bf2d32b458f9b959381381be0047d159a75de0d#diff-b73532e4e80fb3619a11139f58fdf38bf2ca17a74e29344b7f56612608dfa670R586-R611).

However, I realize I don't actually know what is being tested for in these [waypoint assertions](https://github.com/osrf/rmf_fleet_adapter_python/pull/6/commits/2bf2d32b458f9b959381381be0047d159a75de0d#diff-b73532e4e80fb3619a11139f58fdf38bf2ca17a74e29344b7f56612608dfa670R643-R647). Would hope to get your input on this. On visual inspection of the Loop test, i think the mock robot is behaving as expected.

Reply:
Nice work! It also took a while for me to interpret the assertion of this `visited_waypoints`. From what I understand, the dictionary is storing the "[count](https://github.com/osrf/rmf_fleet_adapter_python/pull/6/commits/2bf2d32b458f9b959381381be0047d159a75de0d#diff-b73532e4e80fb3619a11139f58fdf38bf2ca17a74e29344b7f56612608dfa670R586-R611)"  each waypoint is being visited. Other than that, the route looks logical.

Would wish to print out the waypoint [here](https://github.com/osrf/rmf_fleet_adapter_python/blob/efa4c1f181ee8ae82fb59e532ab83eb9bec18cea/rmf_fleet_adapter_python/scripts/rmf_fleet_adapter_python/test_adapter.py#L320), which will help in debugging.

Reply:
I'm not sure i understand correctly, but do you mean to print out the docking waypoint? I have modified the code to print out the "trace" sequence of waypoints traversed, rather than counts. If so, the docking waypoints are recorded in in [`self.dockings`](https://github.com/osrf/rmf_fleet_adapter_python/blob/efa4c1f181ee8ae82fb59e532ab83eb9bec18cea/rmf_fleet_adapter_python/scripts/rmf_fleet_adapter_python/test_adapter.py#L313) and I have modified the code to [print](https://github.com/cnboonhan/rmf_fleet_adapter_python/blob/update/release-1.1/rmf_fleet_adapter_python/scripts/rmf_fleet_adapter_python/test_delivery.py#L193-L195) out these docking waypoints.